src/core/electrostatics_magnetostatics/p3m-dipolar.hpp

```diff
@@ -268,19 +270,23 @@ inline double dp3m_add_pair_force(Particle *p1, Particle *p2,
     auto const mjxr = vector_product(dip2, d);
 
     // Calculate real-space torques
-    p1->f.torque += dipole.prefactor * (-mixmj * B_r + mixr * (mjr * C_r));
-    p2->f.torque += dipole.prefactor * (mixmj * B_r + mjxr * (mir * C_r));
+    auto const torque1 = dipole.prefactor * (-mixmj * B_r + mixr * (mjr * C_r));
+    auto const torque2 = dipole.prefactor * (mixmj * B_r + mjxr * (mir * C_r));
 #endif
 #ifdef NPT
 #if USE_ERFC_APPROXIMATION
     auto const fac = dipole.prefactor * p1->p.dipm * p2->p.dipm * exp_adist2;
 #else
     auto const fac = dipole.prefactor * p1->p.dipm * p2->p.dipm;
 #endif
-    return fac * (mimj * B_r - mir * mjr * C_r);
+    auto const _energy = fac * (mimj * B_r - mir * mjr * C_r);
+#else
+    auto const _energy = 0.0;
 #endif
+    return std::make_tuple(_energy, force, torque1, torque2);
   }
-  return 0.0;
+  return std::make_tuple(0.0, Utils::Vector3d{}, Utils::Vector3d{},
```

Comment:
why is there sometimes an energy returned and sometimes not and why is it ever returned if it's not used anywhere in the code?

Reply:
I don't know why NPT requires energies during the calculation of forces. More worrisome here is the code duplication of the energy calculation in `dp3m_add_pair_force()` and `dp3m_add_pair_energy()`, the latter containing comments ([1](https://github.com/espressomd/espresso/blob/54bbb885120e166c0a6945ef196bcacabf10a971/src/core/electrostatics_magnetostatics/p3m-dipolar.hpp#L311), [2](https://github.com/espressomd/espresso/blob/54bbb885120e166c0a6945ef196bcacabf10a971/src/core/electrostatics_magnetostatics/p3m-dipolar.hpp#L336-L337)) suggesting the energy formula used in the former ([3](https://github.com/espressomd/espresso/blob/54bbb885120e166c0a6945ef196bcacabf10a971/src/core/electrostatics_magnetostatics/p3m-dipolar.hpp#L280-L282)) is wrong.